File: chapter-2/ruby/2.3.rb
# Implement an algorithm to delete a node in the middle of a single linked list, given only access to that node.

# Complexity O(1)
def delete_node node
  if !node || !node[:node_next] then return false end

  node_next = node[:node_next]
  node[:data] = node_next[:data]
  node[:node_next] = node_next[:node_next]

  true
end

# testing
require 'test/unit/assertions'
include Test::Unit::Assertions

class LinkedList
  attr_reader :size

  def initialize
    @size = 0
    @head = nil
    @tail = nil
  end
  
  private
  def create_node data
    {
      data: data,
      node_next: nil
    }
  end

  public
  def add_node data
    node = create_node data
    @size = @size + 1

    if !@head
      @head = node
    else
      @tail[:node_next] = node 
    end

    @tail = node
  end
end

linked = LinkedList.new
a = linked.add_node 'a'
b = linked.add_node 'b'
c = linked.add_node 'c'

delete_node(b)
assert_equal 'c', a[:node_next][:data]

File: chapter-2/js/2.3.js
// Implement an algorithm to delete a node in the middle of a single linked list, given only access to that node.

// Complexity: O(1)
function deleteNode(node) {
  if (!node || !node.nodeNext) { return false; }

  var nodeNext = node.nodeNext;
  node.data = nodeNext.data;
  node.nodeNext = nodeNext.nodeNext;

  return true;
}

// Testing
var assert = require('assert');

function LinkedList() {
  this.size = 0;
  this.head = null;
  this.tail = null;
};

LinkedList.prototype = {
  _createNode: function(data) {
    return {
      data: data,
      nodeNext: null
    };
  },

  addNode: function(data) {
    var node = this._createNode(data);
    
    if (!this.head) {
      this.head = node;
    } else {
      this.tail.nodeNext = node;
    }

    this.tail = node;

    return this.tail;
  }
};

var linked = new LinkedList();
var a = linked.addNode('a');
var b = linked.addNode('b');
var c = linked.addNode('c');

deleteNode(b);
assert.equal('c', a.nodeNext.data);
